librad/src/id/entity/data.rs

```diff
@@ -0,0 +1,209 @@
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::id::entity::{Entity, Error};
+use multihash::{Multihash, Sha2_256};
+use olpc_cjson::CanonicalFormatter;
+use serde::{de::DeserializeOwned, Deserialize, Serialize, Serializer};
+use std::collections::{BTreeSet, HashMap, HashSet};
+
+#[derive(Clone, Deserialize, Serialize, Debug, PartialEq)]
+pub struct EntitySignatureData {
+    pub user: Option<String>,
+    pub sig: String,
+}
+
+fn ordered_set<S>(value: &HashSet<String>, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    let ordered: BTreeSet<String> = value.iter().cloned().collect();
+    ordered.serialize(serializer)
+}
+
+#[derive(Clone, Deserialize, Serialize, Debug, Default)]
+pub struct EntityData<T> {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub revision: Option<u64>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub hash: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub parent_hash: Option<String>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub signatures: Option<HashMap<String, EntitySignatureData>>,
+
+    #[serde(
+        skip_serializing_if = "HashSet::is_empty",
+        serialize_with = "ordered_set",
+        default
+    )]
+    pub keys: HashSet<String>,
+    #[serde(
+        skip_serializing_if = "HashSet::is_empty",
+        serialize_with = "ordered_set",
+        default
+    )]
+    pub certifiers: HashSet<String>,
+
+    pub info: T,
+}
+
+impl<T> EntityData<T>
+where
+    T: Serialize + DeserializeOwned + Clone + Default,
+{
+    pub fn to_json_writer<W>(&self, writer: W) -> Result<(), Error>
+    where
+        W: std::io::Write,
+    {
+        serde_json::to_writer(writer, self).map_err(Error::SerializationFailed)?;
+        Ok(())
+    }
+    pub fn to_json_string(&self) -> Result<String, Error> {
+        Ok(serde_json::to_string(self).map_err(Error::SerializationFailed)?)
+    }
+
+    pub fn from_json_reader<R>(r: R) -> Result<Self, Error>
+    where
+        R: std::io::Read,
+    {
+        Ok(serde_json::from_reader(r).map_err(Error::SerializationFailed)?)
+    }
+    pub fn from_json_str(s: &str) -> Result<Self, Error> {
+        Ok(serde_json::from_str(s).map_err(Error::SerializationFailed)?)
```

Comment:
twice :smile: :man_facepalming:

Reply:
Hahaha, I spotted it a few times :smile: Maybe just do a quick grep. These can be a bit tricky because sometimes it's valid if `?` is calling `into()` to convert the error, so it's not always obvious.

Reply:
Hopefully i fixed them all...

Reply:
:crossed_fingers: